### ISSUE TITLE
Array assignment memory leak

### ISSUE THREAD

**Issue by akkes** [numpy/numpy#17057]
When assigning enough elements of a numpy array a memory leak of 270-290kB happen. The leak happens in regular bursts. It happens too when assigning an array `storage[n] = np.ones(item_size)` with a 2D storage `storage = np.zeros(item_nb, item_size)`

![ram_usage](https://user-images.githubusercontent.com/1679348/89904441-3b7afc00-dbe9-11ea-8b1f-3135e59bfaac.png)

### Reproducing code example:

```python
"""Minimal leak example"""
import numpy as np
import psutil
import os

# may need to add a zero to reproduce in 
item_nb = 30000
# allocate space for items
storage = np.zeros(item_nb)
process = psutil.Process(os.getpid())
old_mem_usage = process.memory_info().rss
for n in range(item_nb):
    storage[n] = 1.0

mem_usage = process.memory_info().rss
print("{}B leaked".format(mem_usage - old_mem_usage))
```

### Error message:

None

### Numpy/Python version information:

Confirmed on versions:
1.19.1 3.8.5 (default, Aug 10 2020, 14:21:14) 
[GCC 9.3.0]
1.18.1 3.7.6 (default, Jan  8 2020, 19:59:22) 
[GCC 7.3.0]
1.19.1 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]

If you have more questions i would be happy to answer them :-)

**Comment by seberg** [comment]
@akkes It sounds like you are simply measuring the behaviour of your operating system. The OS does typically not actually allocate all of the memory immediately, but rather do it on the fly as needed  – it knows that everything untouched is 0 (if you use `np.zeros` or `np.empty`). So ever time you store enough values, more memory is required.
The size of the "leak" is probably the page size of the operating system and should match almost exactly (or even exactly) with the number of new elements stored.

In short, no this is not a leak, its just curious operating system memory management.

**Comment by akkes** [comment]
Indeed, it does not happen when initializing with `storage = np.random.rand(item_nb)`

That's my bad, i should have checked better. Thanks for your fast and explanatory answer! :slightly_smiling_face:

**Comment by seberg** [comment]
No worries, this comes up once every few months, its a curious behaviour.  To proof leaks, you usually want to repeat a certain function in an infinite loop and show that memory keeps increasing forever.
